<launch>

  <!-- NOTE: pcap replay node does not implement clock -->
  <param name="use_sim_time" value="false"/>

  <arg name="loop" default="false" doc="request loop playback"/>
  <arg name="play_delay" default="0" doc="playback start delay in seconds"/>
  <arg name="progress_update_freq" default="3.0"
    doc="playback progress update frequency per second"/>

  <arg name="ouster_ns" default="ouster" doc="Override the default namespace of all ouster nodes"/>
  <arg name="metadata" doc="path to read metadata file when replaying sensor data"/>
  <arg name="pcap_file" doc="file name to use for the recorded pcap file"/>
  <arg name="timestamp_mode" default="TIME_FROM_INTERNAL_OSC"
    doc="A parameter that allows you to override the timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="ptp_utc_tai_offset" default="-37.0"
    doc="UTC/TAI offset in seconds to apply when using TIME_FROM_PTP_1588"/>
  <arg name="viz" default="true" doc="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find ouster_ros)/config/viz.rviz" doc="optional rviz config file"/>

  <arg name="tf_prefix" default=" " doc="namespace for tf transforms"/>
  <arg name="sensor_frame" default="os_sensor"
    doc="sets name of choice for the sensor_frame tf frame, value can not be empty"/>
  <arg name="lidar_frame" default="os_lidar"
    doc="sets name of choice for the os_lidar tf frame, value can not be empty"/>
  <arg name="imu_frame" default="os_imu"
    doc="sets name of choice for the os_imu tf frame, value can not be empty"/>
  <arg name="point_cloud_frame" default=" "
    doc="which frame to be used when publishing PointCloud2 or LaserScan messages.
    Choose between the value of sensor_frame or lidar_frame, leaving this value empty
    would set lidar_frame to be the frame used when publishing these messages."/>
  <arg name="pub_static_tf" default="true" doc="
    which frame to be used when publishing PointCloud2 or LaserScan messages.
    Choose between the value of sensor_frame or lidar_frame, leaving this value empty
    would set lidar_frame to be the frame used when publishing these messages."/>

  <arg name="dynamic_transforms_broadcast" default="false"
    doc="static or dynamic transforms broadcast"/>
  <arg name="dynamic_transforms_broadcast_rate" default="1.0"
    doc="set the rate (Hz) of broadcast when using dynamic broadcast; minimum value is 1 Hz"/>

  <arg name="proc_mask" default="IMG|PCL|IMU|SCAN" doc="
    The IMG flag here is not supported and does not affect anything,
    to disable image topics you would need to omit the os_image node
    from the launch file"/>

  <arg name="scan_ring" default="0" doc="
    use this parameter in conjunction with the SCAN flag
    and choose a value the range [0, sensor_beams_count)"/>

  <arg name="point_type" default="original"
    doc="point type for the generated point cloud; available options: {
    original,
    native,
    xyz,
    xyzi,
    xyzir
    }"/>

  <arg name="organized" default="true"
    doc="generate an organzied point cloud"/>
  <arg name="destagger" default="true"
    doc="enable or disable point cloud destaggering"/>

  <arg name="min_range" default="0.0"
    doc="minimum lidar range to consider (meters)"/>
  <arg name="max_range" default="10000.0"
    doc="maximum lidar range to consider (meters)"/>

  <arg name="min_scan_valid_columns_ratio" default="0.0"
    doc="The minimum ratio of valid columns for processing the LidarScan [0, 1]"/>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_nodelet_mgr"
      output="screen" required="true" args="manager"/>
  </group>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet"
      name="os_node" output="screen" required="true"
      launch-prefix="bash -c 'sleep $(arg play_delay); $0 $@' "
      args="load ouster_ros/OusterPcap os_nodelet_mgr">
      <param name="~/metadata" value="$(arg metadata)"/>
      <param name="~/pcap_file" value="$(arg pcap_file)"/>
      <param name="~/loop" value="$(arg loop)"/>
      <param name="~/progress_update_freq" value="$(arg progress_update_freq)"/>
    </node>
  </group>

  <include file="$(find ouster_ros)/launch/common.launch">
    <arg name="ouster_ns" value="$(arg ouster_ns)"/>
    <arg name="viz" value="$(arg viz)"/>
    <arg name="rviz_config" value="$(arg rviz_config)"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="sensor_frame" value="$(arg sensor_frame)"/>
    <arg name="lidar_frame" value="$(arg lidar_frame)"/>
    <arg name="imu_frame" value="$(arg imu_frame)"/>
    <arg name="point_cloud_frame" value="$(arg point_cloud_frame)"/>
    <arg name="pub_static_tf" value="$(arg pub_static_tf)"/>
    <arg name="timestamp_mode" value="$(arg timestamp_mode)"/>
    <arg name="ptp_utc_tai_offset" value="$(arg ptp_utc_tai_offset)"/>
    <arg name="dynamic_transforms_broadcast"
      value="$(arg dynamic_transforms_broadcast)"/>
    <arg name="dynamic_transforms_broadcast_rate"
      value="$(arg dynamic_transforms_broadcast_rate)"/>
    <arg name="proc_mask" value="$(arg proc_mask)"/>
    <arg name="scan_ring" value="$(arg scan_ring)"/>
    <arg name="point_type" value="$(arg point_type)"/>
    <arg name="organized" value="$(arg organized)"/>
    <arg name="destagger" value="$(arg destagger)"/>
    <arg name="min_range" value="$(arg min_range)"/>
    <arg name="max_range" value="$(arg max_range)"/>
    <arg name="min_scan_valid_columns_ratio"
        value="$(arg min_scan_valid_columns_ratio)"/>
  </include>


</launch>
